### ISSUE TITLE
fed, stl: Some discussion following the ETC

### ISSUE THREAD

**Issue by Jollyfant** [EIDA/mediatorws#30]
Hey, I wanted to open some points for discussion and thought GitHub was a good place for it so everyone can follow and join. After going through the ETC minutes again I have one burning question:

- What problems does the StationLite webservice solve?

Possibly I am entirely overlooking the problem that you are solving, but I think that **1)** virtual networks can be resolved by the routing service when all nodes participate (see recent communication on the ETC mailing list about this). **2)** Expansion to the stream level can be done at request-time going through `routing`->`station`->`dataselect`. Naturally it will be slower than a local cache, but adhere more closely to the philosophy of EIDA, that is a federated archive. There will be some unnecessary hits on the endpoints (e.g. when querying for a minimum latitude) but that is entirely acceptable IMO. Keeping cache state in sync is extremely difficult and adds a layer of complexity that can be avoided. There may be a mismatch between a data center inventory, and the "same" Federator inventory.

With all things considered, I am under the impression that we could support federation in EIDA without a local cache, despite being marginally slower. Please correct me here if I am wrong. I have some experience and know as much as you there are some tricky bits without caching that will need to be solved. But here are some ideas to consider:

- The plan is to query routing, then station to resolve requests to stream level, and finally query dataselect for each stream (each datacenter in parallel). mSEED per request is buffered in memory (to prevent record mixing) and then forwarded to the user when completed.

- For station requests we can query the routing service and then directly the station webservice. We should definitely not split inventory requests by stream as it would take forever to finish. :bulb: Did you know: there are 41474 channels in EIDA.

One thing I noticed is that Z3 queries are really slow because of the excessive number of routes returned. But nothing that this particular optimization cannot fix.

- Bundle similar routing queries to the same datacenter. E.g. the routing service returns one route for every Z3 station. But we can pass multiple stations to the endpoints if we combine these routes (e.g. `station=A001A,A002A,A003A`). Usually this means each Z3 endpoint is hit only once. This can also be applied for networks in certain cases. E.g. `network=N*` can be one query to ORFEUS Data center (despite the routing service returning multiple routes for NA, NL, etc..).

This is a pretty late-stage optimization and can be implemented in the routing service too. It doesn't have to be the job of the Federator.

- Optimize exceptional queries e.g. `network=*&station=*` to hit all the endpoints returned by the routing service only once. This could also be hoisted to the routing service, that when the network & station parameters are both `*` only one route per data center is returned. This will make the query from @javiquinte reported in issue #28 almost instant :zap:.

A very specific optimization but really useful nonetheless. A query for all stations within EIDA can be expected often.

To conclude, I hope these are some points to consider and to think about an alternative federated architecture of the system. Also I want to emphasize that the code right now looks really professional :+1: for that @damb and co.

**Comment by damb** [comment]
> 1) virtual networks can be resolved by the routing service when all nodes participate (see recent communication on the ETC mailing list about this). 

Virtual networks are are not part of this discussion. Right, VNets are handled both from `eidaws-routing` and `stationlite`.

**Comment by damb** [comment]
> One thing I noticed is that Z3 queries are really slow because of the excessive number of routes returned. But nothing that this particular optimization cannot fix.
>
>    Bundle similar routing queries to the same datacenter. E.g. the routing service returns one route for every Z3 station. But we can pass multiple stations to the endpoints if we combine these routes (e.g. station=A001A,A002A,A003A). Usually this means each Z3 endpoint is hit only once. This can also be applied for networks in certain cases. E.g. network=N* can be one query to ORFEUS Data center (despite the routing service returning multiple routes for NA, NL, etc..).
>
> This is a pretty late-stage optimization and can be implemented in the routing service too. It doesn't have to be the job of the Federator.

Mathijs, your suggestion is worth considering but unfortunately is does not meet the federator *registered-mode* ETC requirement. Correct me please if I am wrong ;).

**Comment by damb** [comment]
>    Optimize exceptional queries e.g. network=*&station=* to hit all the endpoints returned by the routing service only once. This could also be hoisted to the routing service, that when the network & station parameters are both * only one route per data center is returned. This will make the query from @javiquinte reported in issue #28 almost instant zap.

> A very specific optimization but really useful nonetheless. A query for all stations within EIDA can be expected often.

That is why I labled #28 as an *enhancement*.

I suggest an implementation using Apache's `mod_cache` (together with `stationlite`'s `mod_wsgi` implementation (already implemented)). But of course I'm open for additional suggestions.

**Comment by damb** [comment]
> To conclude, I hope these are some points to consider and to think about an alternative federated architecture of the system. Also I want to emphasize that the code right now looks really professional +1 for that @damb and co.

Thx :). :+1:

**Comment by javiquinte** [comment]
Hi @damb ! Thanks for the long answer. It's good for the participation of the other ETC members.
I would like to comment just one thing that, at least from the point of view of the previous discussions, is wrong.

> Second, we have to go the path making granular endoint requests since earlier or later we have to implement the registered mode for mediator. Returning detailed logging information is only possible if the federator implements a registered mode, too. This is a requirement by ETC.

There is no such requirement for the Federator. Actually, the specifications say exactly the opposite.

> As the pre-existing FDSN service specification does not describe infrastructure for data tracing and citation, and error tracking, these features are added to the mediator only.
[...]
Query: follows the respective FDSN specification. Data returned is based on best-effort. Failure, and failure reasons, of individual node endpoints are opaque to the user
Comment from Philipp: [...]Current FDSN service specification does not allow to give detailed feedback. if at our endpoint, a part of the requested stations is available, we take a best effort approach and return what we have. I propose for the federator to do the same.

My opinion is that his is not a good idea, but you are free to split a big request in thousands of small requests if you want. This is a technical decision and we can measure if it has or not an impact in the performance. But, this was never a requirement for the Federator.

**Comment by javiquinte** [comment]
Hi @Jollyfant !
Could you explain why this *definitely* has to be granular in the Federator?

> Granular endpoint requests are definitely required for logging purposes in case of dataselect.

We have no chance to inform the user anything in a granular way. The code for the Federator is either 200 or 204 (all or nothing). Apart from that, POST requests are usually faster than many small GET requests, despite that it's less critical for dataselect if compared with station or wfcatalog.

**Comment by Jollyfant** [comment]
@javiquinte Ok it not a requirement but a good idea :bear: 

> My opinion is that his is not a good idea, but you are free to split a big request in thousands of small requests if you want. This is a technical decision and we can measure if it has or not an impact in the performance. But, this was never a requirement for the Federator.

From experience I know that splitting dataselect to the stream level solves a lot of problems. If we don't, some requests will hit `413 - Payload Too Large` and the federator is unable to give feedback. This will frustrate users as the data is just not in their result, despite the 200 status code from earlier streamed data.

Individual stream requests will never return 413 if split (where necessary) by time window too (e.g. 7 days) so that it will never exceed the number of maximum samples configured by the service.

It also allows us to have better control over the requests that are sent to the endpoints. We can do throttling when necessary.

**Comment by Jollyfant** [comment]
And since the Federator will need to buffer data to memory it becomes easier to flush when a small single channel request is completed.

**Comment by Jollyfant** [comment]
> We have no chance to inform the user anything in a granular way. The code for the Federator is either 200 or 204 (all or nothing). Apart from that, POST requests are usually faster than many small GET requests, despite that it's less critical for dataselect if compared with station or wfcatalog.

The Federator response (when streaming) is either 200 (all), 200 (some), or 204 (nothing).. that is an inherent complication of HTTP we need to deal with.

**Comment by damb** [comment]
Hi @Jollyfant,
> In case of the routing service, there will only be a mismatch with routes, meaning in the worst case data is not found at the specified data center. With a station cache there may be an inconsistency in station latitude, longitude, including other cached attributes.

One question: Where does the current `eidaws-routing` implementation get the geolocation data from? Within the `localconfig` files there is no such data available. I guess, it is also `fdsnws-station`.

Respective the point: `stationlite` vs `eidaws-routing` -> `fdsnws-station`: I guess, it is a matter of taste which approach to choose. But, the new federator implementation is easily extensible to implement a comparison testing both approaches.

> Resolving requests to the channel level is not feasible for inventories. You cannot make over 40000 requests to the endpoints to collect information on all channels in EIDA. Even at 130 requests per second it will take you 5 minutes. An inventory request network=NL should be a single inventory request. Another argument is that StationXML will include much redudant information when queried per channel, as network & station details are repeated N times. For logging purposes, you can resolve all available channels by querying fdsnws-station with a second request for this granular information. This approach seems a lot more practical.

I was told that detailed logging is a requirement by ETC for the **mediator**. As I told already this is only feasible if the federator also implements a registered mode. But I cannot make two requests since the results may differ. And yes, I suggest doing it also at request time.

> I am not up to date as to what requirement this violates. Do you mean the granular logging?
Yes.

> For the sake of argument, let us pretend to agree that we need a local cache. Can we then not use SeisComP3, harvest all EIDA channels and stick them in the database and start fdsnws. That will resolve requests to the stream level too, right? Then we get to re-use a good part of our infrastructure.

This question has to answer @kaestli.

cheers, Daniel

**Comment by Jollyfant** [comment]
> One question: Where does the current eidaws-routing implementation get the geolocation data from? Within the localconfig files there is no such data available. I guess, it is also fdsnws-station.

This is correct. It was implemented and discussed in the ETC before you joined to tackle queries for latitude, longitude without having to hit the endpoints.

**Comment by damb** [comment]
@Jollyfant:

> @javiquinte Ok it not a requirement but a good idea bear
>
 >   My opinion is that his is not a good idea, but you are free to split a big request in thousands of small requests if you want. This is a technical decision and we can measure if it has or not an impact in the performance. But, this was never a requirement for the Federator.
>
> From experience I know that splitting dataselect to the stream level solves a lot of problems. If we don't, some requests will hit 413 - Payload Too Large and the federator is unable to give feedback. This will frustrate users as the data is just not in their result, despite the 200 status code from earlier streamed data.
>
> Individual stream requests will never return 413 if split (where necessary) by time window too (e.g. 7 days) so that it will never exceed the number of maximum samples configured by the service.
>
> It also allows us to have better control over the requests that are sent to the endpoints. We can do throttling when necessary.

Currently I'm implementing the HTTP status code 413 handling (for `fdsnws-dataselect` and `eidaws-wfcatalog`). See also #29.

**Comment by damb** [comment]
Hi @javiquinte,

>  We have no chance to inform the user anything in a granular way. The code for the Federator is either 200 or 204 (all or nothing). Apart from that, POST requests are usually faster than many small GET requests, despite that it's less critical for dataselect if compared with station or wfcatalog.

That's wrong. If the federator logs the endpoint HTTP status codes (including e.g. the message), to e.g. a DB, then there is the chance to log in a granular way. However, the federator will perform slightly slower depending on the implementation. But I guess it would be marginal.

Endpoint requests are made via POST by the federator.

cheers, Daniel

**Comment by damb** [comment]
> This is correct. It was implemented and discussed in the ETC before you joined to tackle queries for latitude, longitude without having to hit the endpoints.

Ok, good to know.

**Comment by damb** [comment]
@Jollyfant:

> The Federator will always have to wait for the weakest link, and I agree that the endpoints need improvements. I believe that GFZ is working on putting these services in containers so we can easily deploy more than a single entity.

Simply putting federator services into containers does not scale well, since you have the overhead of multiple webserver instances (e.g. Apache. nginx). If the application is set up properly (e.g. by means of `mod_wsgi`) apache is able to work with multiple concurrent federator instances. Also, this makes caching much easier.

**Comment by damb** [comment]
Citations (as implemented by [fdsnws_fetch](https://github.com/andres-h/fdsnws_scripts/blob/master/fdsnwsscripts/fdsnws_fetch.py)) are perhaps an additional important point why we won't come around detailed federator logging facilities/ federator registered-mode.

**Comment by damb** [comment]
@Jollyfant 
> And since the Federator will need to buffer data to memory it becomes easier to flush when a small single channel request is completed.

Buffering data in memory is no solution for the federator. Just think about handling multiple requests each requesting several GB data. Even with HTTP status code 413 handling this would easily lead to a overflow on the federator side. That is why the streaming federator version implements the following approach:

- Download data from endpoints is performed streamed, multithreaded. The data is kept in Tempfiles.
- As soon a data for a complete stream epoch is available we deliver data to the client. Also streamed. (Prevents gateway errors - HTTP code 504)

cheers, Daniel

**Comment by Jollyfant** [comment]
> Simply putting federator services into containers does not scale well, since you have the overhead of multiple webserver instances (e.g. Apache. nginx). If the application is set up properly (e.g. by means of mod_wsgi) apache is able to work with multiple concurrent federator instances. Also, this makes caching much easier.

Sorry I meant node services (i.e. dataselect & station) at the end points. We can deploy multiple of those.

> Buffering data in memory is no solution for the federator. Just think about handling multiple requests each requesting several GB data. Even with HTTP status code 413 handling this would easily lead to a overflow on the federator side. That is why the streaming federator version implements the following approach.

> * Download data from endpoints is performed streamed, multithreaded. The data is kept in Tempfiles.
> * As soon a data for a complete stream epoch is available we deliver data to the client. Also streamed. (Prevents gateway errors - HTTP code 504)


If we take a single stream @ 200Hz for one week (assuming the Federator splits request by channel & week length) we get to roughly 120MB that needs to be buffered in memory before it is automatically flushed and freed.

    http://www.orfeus-eu.org/fdsnws/dataselect/1/query?sta=G100&cha=HGZ&start=2017-06-01&end=2017-06-07

It would be faster than disk I/O, so I would say keep this possibility in mind. Ultimately, I don't know the performance difference between memory and disk in this case. If there is barely a difference, tmpfiles are definitely the desired solution!

**Comment by javiquinte** [comment]
There is always the chance of sending data as soon as the first records are received from the endpoints. That avoids disk I/O, memory consumption, user waiting time, etc. But again, it's an implementation strategy. The most important thing is that it should perform well.

**Comment by Jollyfant** [comment]
> There is always the chance of sending data as soon as the first records are received from the endpoints. That avoids disk I/O, memory consumption, user waiting time, etc. But again, it's an implementation strategy. The most important thing is that it should perform well.

Then it is impossible to do parallel requests because records would get mixed. So I would buffer a full request to memory to guarantee integrity of the mSEED, and then flush when complete. When the Federator is nearly out of resources it can go in to emergency mode and fall back to tmpfiles, perhaps.

**Comment by javiquinte** [comment]
But mSEED files could formally have some interlaced records (mixed channels) and still be chronologically ordered. F.i., [BHZ-rec1, BHZ-rec2, BHE-rec1, BHZ-rec3, BHE-rec2, BHE-rec3]. :-)
Again, it was just a comment, my personal opinion is that any implementation which performs well on a reasonable hardware should be OK.

**Comment by kaestli** [comment]
Note that modern OS cache disk access in a fairly clever way, and transparently/adaptively use the entire RAM not claimed by other processes for this task. If you cache to file, and read again, you can be pretty sure to read actually the OS' RAM cache if the number and size of parallel IO operations allow. Leaving the cache management to the OS will allow for less/simpler code with (i expect) fairly good performance and resource usage.

**Comment by kaestli** [comment]
@javiquinte while multiplexed mseed is formally fully valid, it is quite hard (inefficient) to read, if multiplexing channels are many, and records of different streams are, overall, not time-sorted (as different streams are shipped at different speed from different endpoints).
There are two other options for a "tidier" response:
1. fully time-sort the records (as for RT replay)
2. stream-sort, then time sort (so within a response, a stream interval is always coherent)
There are two reasons we have chosen 2:
a) this is how the existing dataselect endpoints already behave, thus although not specified, the behaviour must be expected by the users.
b) full time-sorting would limit return speed to the slowest stream / endpoint, thus remain below best effort and increase the need for buffering.

**Comment by damb** [comment]
@Jollyfant 
>>   Simply putting federator services into containers does not scale well, since you have the overhead of multiple webserver instances (e.g. Apache. nginx). If the application is set up properly (e.g. by means of mod_wsgi) apache is able to work with multiple concurrent federator instances. Also, this makes caching much easier.
>
> Sorry I meant node services (i.e. dataselect & station) at the end points. We can deploy multiple of those.

This practice by far is not ideal. Especially not for endpoints. Every single container ships its own webserver which simply consumes resources. Correct me if I'm wrong. Do you configure your containers differently? In my opinion you should configure the webserver running the webservice to process requests concurrently. If the load for one webserver gets too big you still can containerize ...

**Comment by Jollyfant** [comment]
> This practice by far is not ideal. Especially not for endpoints. Every single container ships its own webserver which simply consumes resources. Correct me if I'm wrong. Do you configure your containers differently? In my opinion you should configure the webserver running the webservice to process requests concurrently. If the load for one webserver gets too big you still can containerize ...

Pretty sure this must be a communication issue. Surely it can't be a bad idea to add more resources to our end points? I mean, at ODC I would like to run 2 FDSNWS-Dataselect webservices, because with only one during high load we see it struggles to keep up. One instance of SeisComP3 can't utilize all available resources.

**Comment by damb** [comment]
Hi @Jollyfant 

> Pretty sure this must be a communication issue. Surely it can't be a bad idea to add more resources to our end points? I mean, at ODC I would like to run 2 FDSNWS-Dataselect webservices, because with only one during high load we see it struggles to keep up. One instance of SeisComP3 can't utilize all available resources.

Unfortunately I do not know how FDSN webservices from SeisComp3 exactly are implemented and how you guys deploy them. However, when having a look at the code https://github.com/SeisComP3/seiscomp3/blob/master/src/trunk/apps/fdsnws/fdsnws.py it turns out that the implementation is based on the [twisted](https://twistedmatrix.com/trac/) framework, an async networking library. In addition endpoint resources i.e. `fdsnws-dataselect`, `fdsnws-station` and `fdsnws-event` may be enabled/disabled if desired.
Bearing these facts in mind, Mathijs, in deed an deployment approach as follows seems to make sense:
- Run several SeisComp3 FDSN webservice implementations in parallel. Deploy FDSN webservice instances depending on the load for each endpoint resource.
- It is a matter of taste if you deploy the FDSN webservice i.e. twisted instances containerized or not. But of course, when deploying multiple instances wrapping them up into containers might be reasonable.
- Use Apache2 or another webserver as loadbalancer / reverse proxy redirecting incoming requests to the webservices.

----
To clearify the communication issue ;): The federator is implemented using flask, which, in contrast to twisted, is implementing the *multithreaded model* approach. As a WSGI application it perfectly may be deployed with Apache2 + `mod_wsgi` / nginx + `uWSGI`.
Apache2 worker threads run separate python interpreters which themselves run the federator webservice implementation. In this case Apache2 no longer is configured as a reverse proxy. However, most of the configuration is done at the webserver itself. Of course, still you might wrap the bundle Apache2 + `mod_wsgi` + flask (federator) up into a container.

I guess it currently is a matter of taste which implementation/deployment approaches to follow. Future and performance tests will show which way to continue.

cheers, Daniel

**Comment by damb** [comment]
I close this issue. The discussion continued during our most recent ETC meeting at Grenoble. If there are further issues or questions feel free to reopen.
